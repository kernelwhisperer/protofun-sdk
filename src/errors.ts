export class IndexerError extends Error {
  name = "IndexerError"
}
