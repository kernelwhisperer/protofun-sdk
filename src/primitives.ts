export type Timeframe = "Block" | "Minute" | "Hour" | "Day" | "Week"

export enum PriceUnit {
  ETH = "ETH",
  USD = "USD",
  GWEI = "Gwei",
}

export type Block = {
  baseFeePerGas: string // BigInt
  burnedFees: string // BigInt
  firstGasPrice: string // BigInt
  gasFees: string // BigInt
  gasUsed: string // BigInt
  id: string // bytes
  lastGasPrice: string // BigInt
  maxGasPrice: string // BigInt
  minGasPrice: string // BigInt
  minerTips: string // BigInt
  number: string // BigInt
  timestamp: string // BigInt
  // txns: Array<Txn>
  txnCount: number
}

export type SimpleBlock = Omit<Block, "txns" | "timestamp"> & {
  timestamp: string
}

export type Candle = {
  close: string
  high: string
  low: string
  open: string
  timestamp: string
}

export function isCandle(value: unknown): value is Candle {
  return typeof value === "object" && value !== null && "open" in value
}

export function isCandleArray(value: unknown[]): value is Candle[] {
  return Array.isArray(value) && value.length > 0 && isCandle(value[0])
}

export type QueryRequest = {
  /**
   * @default 1000
   */
  limit?: number
  priceUnit?: PriceUnit
  since?: string
  timeframe: Timeframe
  until?: string
}
export type QueryResult = Promise<Candle[] | SimpleBlock[]>
export type QueryFn = (request: QueryRequest) => QueryResult
