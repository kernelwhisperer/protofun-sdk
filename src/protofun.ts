export * from "./errors"
export * from "./metric-declarations"
export * from "./primitives"
