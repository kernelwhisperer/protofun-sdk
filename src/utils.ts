import { MetricId, ProtocolId, QueryFn, SubscribeFn, Timeframe } from "./protofun"

export async function wait(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export const allTimeframes: Timeframe[] = ["Block", "Minute", "Hour", "Day", "Week"]

export type MetricFnsResult = {
  query: QueryFn
  subscribe: SubscribeFn
  supportedTimeframes: Timeframe[]
}

export async function loadMetricFns(
  protocolId: ProtocolId,
  metricId: MetricId,
  packagePath = "protofun/dist/metrics"
): Promise<MetricFnsResult> {
  const module = await import(`${packagePath}/${protocolId}/${metricId}`)
  const { default: query, supportedTimeframes = allTimeframes, subscribe } = module
  return { query, subscribe, supportedTimeframes }
}

export function getLowestTimeframe(supportedTimeframes: Timeframe[]) {
  return supportedTimeframes.find((x) => x !== "Block") as Timeframe
}
