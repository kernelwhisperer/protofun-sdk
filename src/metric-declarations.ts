import { PriceUnit, Timeframe } from "./primitives"
import { allTimeframes } from "./utils"

const METRIC_DECLARATIONS = {
  aave: [],
  btc: ["btc_price"],
  comp: ["tvl", "comp_price"],
  eth: ["base_fee", "eth_price", "tx_cost"],
  mkr: [],
} as const

/**
 * Protocols
 */
export type ProtocolId = keyof typeof METRIC_DECLARATIONS
export type Protocol = {
  enabled?: boolean
  id: ProtocolId
  title: string
  wip?: boolean
}

export const PROTOCOL_MAP: Record<ProtocolId, Protocol> = {
  aave: {
    id: "aave",
    title: "Aave",
  },
  btc: {
    enabled: true,
    id: "btc",
    title: "Bitcoin",
  },
  comp: {
    enabled: true,
    id: "comp",
    title: "Compound",
    wip: true,
  },
  eth: {
    enabled: true,
    id: "eth",
    title: "Ethereum",
  },
  mkr: {
    id: "mkr",
    title: "MakerDAO",
  },
} as const

export const PROTOCOL_IDS = Object.keys(PROTOCOL_MAP) as ProtocolId[]

export const PROTOCOLS = Object.values(PROTOCOL_MAP).sort((a, b) => {
  if (a.enabled === b.enabled) {
    return 0
  } else if (a.enabled && !b.enabled) {
    return -1
  } else {
    return 1
  }
})

/**
 * Metrics
 */

// Not working:
// type ValueOf<T> = T[keyof T];
// export type MetricId = ValueOf<typeof METRIC_LIST_DEFINITION>[number];
export type MetricIdForProtocol<T extends ProtocolId> = typeof METRIC_DECLARATIONS[T][number]
export type MetricId = MetricIdForProtocol<ProtocolId>

export type Variant = {
  label: string
  precision: number
  value?: string
}

export type Metric = {
  allowCompactPriceScale?: boolean
  disallowAlerts?: boolean
  disallowCandleType?: boolean
  disallowLiveMode?: boolean
  id: MetricId
  precision: number
  preferredLogScale?: boolean
  priceUnits: PriceUnit[]
  protocol: ProtocolId
  significantDigits: number[]
  timeframes: Timeframe[]
  title: string
  variants?: Variant[]
  wip?: boolean
}

// type MapType<T extends ProtocolId> = Record<
//   T,
//   Record<MetricIdsOfProtocol<T>, Metric>
// >;
type MetricsMapType = Record<ProtocolId, Partial<Record<MetricId, Metric>>>

export const METRICS_MAP: MetricsMapType = {
  aave: {},
  btc: {
    btc_price: {
      id: "btc_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "btc",
      significantDigits: [0],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "BTC price",
    },
  },
  comp: {
    comp_price: {
      id: "comp_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "comp",
      significantDigits: [2],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Comp price",
    },
    tvl: {
      allowCompactPriceScale: true,
      disallowCandleType: true,
      id: "tvl",
      precision: 1,
      priceUnits: [PriceUnit.USD],
      protocol: "comp",
      significantDigits: [0],
      timeframes: ["Hour", "Day"],
      title: "Total value locked",
      variants: [{ label: "V3 USDC - Wrapped Ether Pool", precision: 1 }],
      wip: true,
    },
  },
  eth: {
    base_fee: {
      id: "base_fee",
      precision: 1e9,
      preferredLogScale: true,
      priceUnits: [PriceUnit.GWEI],
      protocol: "eth",
      significantDigits: [2],
      timeframes: allTimeframes,
      title: "Base fee per gas",
    },
    eth_price: {
      id: "eth_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "eth",
      significantDigits: [2],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Ether price",
    },
    tx_cost: {
      id: "tx_cost",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD, PriceUnit.ETH],
      protocol: "eth",
      significantDigits: [2, 5],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Transaction cost",
      variants: [
        { label: "ETH Transfer", precision: 1e18 / 21_000 },
        { label: "ERC20 Approval", precision: 1e18 / 45_000 },
        { label: "ERC20 Transfer", precision: 1e18 / 65_000 },
        { label: "NFT Transfer", precision: 1e18 / 85_000 },
        { label: "Uniswap V2 Swap", precision: 1e18 / 150_000 },
        { label: "Uniswap V3 Swap", precision: 1e18 / 185_000 },
        { label: "L2 Deposit", precision: 1e18 / 250_000 },
      ],
    },
  },
  mkr: {},
}

export const METRICS = PROTOCOL_IDS.map((protocolId) =>
  Object.values(METRICS_MAP[protocolId] || {})
).flat()

export function getMetric(protocolId: ProtocolId, metricId: MetricId): Metric {
  const metric = METRICS_MAP[protocolId][metricId]

  if (!metric) {
    throw new Error("Metric not found.")
  }

  return metric
}

export function getMetricPrecision(metric: Metric, variantIndex: number): number {
  const { variants, precision: defaultPrecision } = metric
  return variants ? variants[variantIndex].precision : defaultPrecision
}

export function getSignificantDigits(metric: Metric, priceUnitIndex: number): number {
  return metric.significantDigits[priceUnitIndex]
}
