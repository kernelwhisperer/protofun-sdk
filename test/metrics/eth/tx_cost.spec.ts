import { expect, it } from "vitest"

import query, { subscribe } from "../../../src/metrics/eth/tx_cost"
import { Candle, PriceUnit } from "../../../src/primitives"
import { wait } from "../../../src/utils"

const mapToSimpleTransfer = (x: Candle) => ({
  close: (parseFloat(x.close) * 21_000) / 1e18,
  high: (parseFloat(x.high) * 21_000) / 1e18,
  low: (parseFloat(x.low) * 21_000) / 1e18,
  open: (parseFloat(x.open) * 21_000) / 1e18,
  timestamp: x.timestamp,
})

it("Daily candles in USD", async () => {
  // act
  const candles = await query({
    since: "1628121600",
    timeframe: "Day",
    until: "1628553600",
  })
  // assert
  expect(candles.map(mapToSimpleTransfer)).toMatchSnapshot()
})

it("Daily candles in ETH", async () => {
  // act
  const candles = await query({
    priceUnit: PriceUnit.ETH,
    since: "1628121600",
    timeframe: "Day",
    until: "1628553600",
  })
  // assert
  expect(candles.map(mapToSimpleTransfer)).toMatchSnapshot()
})

// TODO
it.skip("Weekly candles in USD", async () => {
  // act
  const candles = await query({
    since: "1627862400",
    timeframe: "Week",
    until: "1630886400",
  })
  // assert
  expect(candles.map(mapToSimpleTransfer)).toMatchSnapshot()
})

// TODO
it.skip("Weekly candles in ETH", async () => {
  // act
  const candles = await query({
    priceUnit: PriceUnit.ETH,
    since: "1627862400",
    timeframe: "Week",
    until: "1630886400",
  })
  // assert
  expect(candles.map(mapToSimpleTransfer)).toMatchSnapshot()
})

it("Match the correct size", async () => {
  // act
  const candles = await query({
    timeframe: "Day",
    until: "1630886400",
  })
  // assert
  expect(candles.length).to.equal(33)
})

it("Block candles", async () => {
  // act
  try {
    await query({
      timeframe: "Block",
    })
  } catch (error) {
    // assert
    expect(error).toMatchInlineSnapshot(
      "[Error: Timeframe 'Block' is not supported for this metric.]"
    )
  }
})

it("Subscribe", async () => {
  // arrange
  const timeframe = "Minute"
  const candles = await query({
    limit: 1,
    timeframe,
  })
  // act
  const unsubscribe = subscribe({
    onNewData: (data) => {
      candles.push(data)
      unsubscribe()
    },
    pollingInterval: 3_000,
    since: candles[0].timestamp,
    timeframe,
  })
  // assert
  await wait(5_000)
  // Sometimes, if we're lucky, we get the next candle + an update to the previous one
  expect([2, 3]).to.include(candles.length)
})
