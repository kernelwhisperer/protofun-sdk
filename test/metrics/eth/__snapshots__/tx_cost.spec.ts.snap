// Vitest Snapshot v1

exports[`Daily candles in ETH 1`] = `
[
  {
    "close": 0.000661136735007,
    "high": 0.009992884178319,
    "low": 0.000021,
    "open": 0.000021,
    "timestamp": "1628121600",
  },
  {
    "close": 0.000900761385468,
    "high": 0.003539734646658,
    "low": 0.000389988036648,
    "open": 0.000727475593362,
    "timestamp": "1628208000",
  },
  {
    "close": 0.001474832091873,
    "high": 0.001765000167315,
    "low": 0.000484205833566,
    "open": 0.000912495468906,
    "timestamp": "1628294400",
  },
  {
    "close": 0.00062745711462,
    "high": 0.016768179150891,
    "low": 0.000340388709717,
    "open": 0.001659142141062,
    "timestamp": "1628380800",
  },
  {
    "close": 0.000780476556279,
    "high": 0.004099623510423,
    "low": 0.000374473406958,
    "open": 0.000565345773006,
    "timestamp": "1628467200",
  },
  {
    "close": 0.000810741847335,
    "high": 0.03919591002894,
    "low": 0.000460962583347,
    "open": 0.000863827320048,
    "timestamp": "1628553600",
  },
]
`;

exports[`Daily candles in USD 1`] = `
[
  {
    "close": 1.8705872211920556,
    "high": 28.429755487317554,
    "low": 0.05314449,
    "open": 0.05721891,
    "timestamp": "1628121600",
  },
  {
    "close": 2.605884672931215,
    "high": 10.43942081727024,
    "low": 1.0611574477192078,
    "open": 2.058384916661845,
    "timestamp": "1628208000",
  },
  {
    "close": 4.665026641482568,
    "high": 5.601616331010962,
    "low": 1.3880147583836242,
    "open": 2.639694267315344,
    "timestamp": "1628294400",
  },
  {
    "close": 1.8912373128895807,
    "high": 53.523860167852554,
    "low": 1.003724611665101,
    "open": 5.248015914971802,
    "timestamp": "1628380800",
  },
  {
    "close": 2.468132232983333,
    "high": 13.077798998249369,
    "low": 1.083164329626015,
    "open": 1.7040200013328446,
    "timestamp": "1628467200",
  },
  {
    "close": 2.54742385109283,
    "high": 126.8026885346238,
    "low": 1.406396841791697,
    "open": 2.7317070490073916,
    "timestamp": "1628553600",
  },
]
`;

exports[`Weekly candles in ETH 1`] = `
[
  {
    "close": 0.00062745711462,
    "high": 0.016768179150891,
    "low": 0.000021,
    "open": 0.000021,
    "timestamp": "1627862400",
  },
  {
    "close": 0.000814385808159,
    "high": 0.03919591002894,
    "low": 0.000363110969844,
    "open": 0.000565345773006,
    "timestamp": "1628467200",
  },
  {
    "close": 0.000500384099544,
    "high": 0.014792705292834,
    "low": 0.000245478205392,
    "open": 0.000914314387665,
    "timestamp": "1629072000",
  },
  {
    "close": 0.001483415978982,
    "high": 0.060672808623651,
    "low": 0.000376643577282,
    "open": 0.000562887430413,
    "timestamp": "1629676800",
  },
  {
    "close": 0.001767709615476,
    "high": 0.051324242827662,
    "low": 0.000794506360095,
    "open": 0.001668661121904,
    "timestamp": "1630281600",
  },
  {
    "close": 0.000875767984728,
    "high": 0.082780327606578,
    "low": 0.000602489908335,
    "open": 0.001721746319832,
    "timestamp": "1630886400",
  },
]
`;

exports[`Weekly candles in USD 1`] = `
[
  {
    "close": 1.889806710668247,
    "high": 53.49032380955078,
    "low": 0.05128872,
    "open": 0.0536655,
    "timestamp": "1627862400",
  },
  {
    "close": 2.6932471623045475,
    "high": 130.56079238819856,
    "low": 1.050563551281756,
    "open": 1.7027310129703912,
    "timestamp": "1628467200",
  },
  {
    "close": 1.6202437143234718,
    "high": 49.33367215160139,
    "low": 0.7241607059064,
    "open": 3.023957690043838,
    "timestamp": "1629072000",
  },
  {
    "close": 4.781746905769108,
    "high": 204.94607352174447,
    "low": 1.1500058345151307,
    "open": 1.822629499677294,
    "timestamp": "1629676800",
  },
  {
    "close": 6.982965616918688,
    "high": 206.72789120068325,
    "low": 2.497133489778585,
    "open": 5.378879066623886,
    "timestamp": "1630281600",
  },
  {
    "close": 2.981298131290905,
    "high": 328.63790059811464,
    "low": 1.810482174546675,
    "open": 6.801380052305952,
    "timestamp": "1630886400",
  },
]
`;
